Add decision on capacity alerts (#59)
vshn/application-catalog-docs: docs/modules/ROOT/pages/explanations/decisions/capacity-alerting.adoc

> = Capacity Alerting and Incident Prevention

Reviewer: ```suggestion
= Handling of Capacity Alerts for Incident Prevention
```

or 

```suggestion
= Alert Service User on Service Instance Capacity
```

From my understanding of the problem description, one of these suggestions would describe it much better?

== Problem

All Databases by VSHN have limited capacity such as storage and memory capacity, or a limit on concurrent connections.
Depending on the database type running out of capacity can lead to downtime or even to data loss.
We need a way to predict capacity issues, so we can react and prevent an incident.

Writing these predictive capacity alerts is generally a solved problem.
We can write Prometheus queries that predict the resource usage and alert based on the predicted usage.
The main question is how we want to handle these alerts.

From experience from similar system, we know that in nearly all cases capacity issues can only be resolved by the customers themselves, either by freeing up resources or by scaling up the instance.

> So first and foremost we need a process to inform the customer about potential capacity issues.

Reviewer: ```suggestion
So first and foremost we need a process to inform the service user about potential capacity issues.
```

Accoring to https://kb.vshn.ch/app-catalog/reference/glossary.html#_personas it's service user, not customer.

== Proposals

=== Alert Responsible Ops

> The simplest approach would be to alert the engineer that does responsible ops.

Reviewer: ```suggestion
The simplest approach would be to alert the engineer that does https://handbook.vshn.ch/role_responsibleops.html[responsible ops^].
```

This is how we are used to handling alerts and how we handled this in previous systems.

In practice this means we will deploy fixed Prometheus alert rules and alert-manager configs in every database namespace that will send alerts to Opsgenie.

The only open question is how we securely distribute the necessary Opsgenie secrets to the database namespace.
This can be solved by https://github.com/vshn/espejo[espejo] or through a custom controller, but we need to be aware that increase the potential attack surface by distributing Opsgenie credentials to many namespaces.

=== Expose User-Workload Monitoring

> OpenShift4 and as a result APPUiO Cloud and Managed come with a versatile user-workload monitoring system that allows a user to specify their own alert-manager configuration on a namespace level.

Reviewer: ```suggestion
APPUiO Cloud and APPUiO Managed OpenShift come with a versatile user-workload monitoring system that allows a user to specify their own alert-manager configuration on a namespace level.
```

For APPUiO Managed Kubernetes Services we would have to figure out how to do it then. Probably this product will provide an OpenShift compatible user workload monitoring stack.

We could expose the alert manager configuration of this system directly in the database claim and the composition will then create an `AlertManagertConfig` resource in the database namespace.
This would enable a customer to decide for themselves how they want to be alerted about capacity issues.

With such a system the database claim could look similar to the following, which would configure all alerts to be sent to the customers slack instance.

[source,yaml]
----
apiVersion: vshn.appcat.vshn.io/v1
kind: VSHNPostgreSQL
metadata:
  name: pgsql-app1-prod
  namespace: prod-app
spec:
  parameters:
    alerting:
      receivers:
      - name: slack
        slackConfigs:
        - apiURL:
            key: url
            name: slack-alert-webhook-secret
          channel: '#alerts'
      route:
        groupBy: [alertname]
        receiver: slack
  writeConnectionSecretToRef:
    name: postgres-creds
----

=== Send E-Mail Alert through Alert-Manager and mailgun

VSHN has the option to send e-mails automatically through https://www.mailgun.com/[mailgun].

> This can be leveraged to infrom customer through e-mail without a lot of setup overhead by the customer.

Reviewer: ```suggestion
This can be leveraged to inform customer through e-mail without a lot of setup overhead by the customer.
```

= Handling of Capacity Alerts for Incident Prevention

== Problem

All Databases by VSHN have limited capacity such as storage and memory capacity, or a limit on concurrent connections.
Depending on the database type running out of capacity can lead to downtime or even to data loss.
We need a way to predict capacity issues, so we can react and prevent an incident.

Writing these predictive capacity alerts is generally a solved problem.
We can write Prometheus queries that predict the resource usage and alert based on the predicted usage.
The main question is how we want to handle these alerts.

> From experience from similar system, we know that in nearly all cases capacity issues can only be resolved by the service users themselves, either by freeing up resources or by scaling up the instance.

Reviewer 2: ```suggestion
From experience from similar systems, we know that in nearly all cases capacity issues can only be resolved by the service users themselves, either by freeing up resources or by scaling up the instance.
```

So first and foremost we need a process to inform the service user about potential capacity issues.

== Proposals

=== Alert Responsible Ops

The simplest approach would be to alert the engineer that does https://handbook.vshn.ch/role_responsibleops.html[responsible ops^].
This is how we are used to handling alerts and how we handled this in previous systems.

> In practice this means we will deploy fixed Prometheus alert rules and alert-manager configs in every database namespace that will send alerts to Opsgenie.

Reviewer 2: ```suggestion
In practice this means we will deploy fixed Prometheus alert rules and Alertmanager configs in every database namespace that will send alerts to Opsgenie.
```

> In practice this means we will deploy fixed Prometheus alert rules and alert-manager configs in every database namespace that will send alerts to Opsgenie.

Reviewer 2: I don't understand the Alertmanager configs part.
Why would we need separate Alertmanager configs for every namespace to alert our ResponsibleOps?

Author: Because on APPUiO Cloud to send alerts to our ResponsibleOps we'll most likely need to use the user-workload monitoring for our alerts. That means we'll need to add Alertmanager config in every namespace we manage and can't depend on cluster-monitoring.

The only open question is how we securely distribute the necessary Opsgenie secrets to the database namespace.
This can be solved by https://github.com/vshn/espejo[espejo] or through a custom controller, but we need to be aware that increase the potential attack surface by distributing Opsgenie credentials to many namespaces.

=== Expose User-Workload Monitoring

> APPUiO Cloud and APPUiO Managed OpenShift come with a versatile user-workload monitoring system that allows a user to specify their own alert-manager configuration on a namespace level.

Reviewer 2: ```suggestion
APPUiO Cloud and APPUiO Managed OpenShift come with a versatile user-workload monitoring system that allows a user to specify their own Alertmanager configuration on a namespace level.
```

> We could expose the Alertmanager configuration of this system directly in the database claim and the composition will then create an `AlertManagertConfig` resource in the database namespace.

Reviewer 3: ```suggestion
We could expose the Alertmanager configuration of this system directly in the database claim and the composition will then create an `AlertManagertConfig` resource in the database namespace.
```
We could expose the Alertmanager configuration of this system directly in the database claim and the composition will then create an `AlertManagerConfig` resource in the database namespace.

> We could expose the alert manager configuration of this system directly in the database claim and the composition will then create an `AlertManagertConfig` resource in the database namespace.

Reviewer 2: ```suggestion
We could expose the Alertmanager configuration of this system directly in the database claim and the composition will then create an `AlertManagertConfig` resource in the database namespace.
```

This would enable a service user to decide for themselves how they want to be alerted about capacity issues.

> With such a system the database claim could look similar to the following, which would configure all alerts to be sent to the service users slack instance.

Reviewer 4: I like this as a first approach, but I feel like this will lead to a lot of duplicated configurations for the customer. We should look into the possibility of having one or multiple such alertmanager configs per user namespace and then distribute them to the instance namespaces.

That should not be too complicated and makes the UX for the user much better IMHO.

Author: We decided in the team that this might be a good idea idea, but we'll start with the proposed, simpler idea for now.

[source,yaml]
----
apiVersion: vshn.appcat.vshn.io/v1
kind: VSHNPostgreSQL
metadata:
  name: pgsql-app1-prod
  namespace: prod-app
spec:
  parameters:
    alerting:
      receivers:
      - name: slack
        slackConfigs:
        - apiURL:
            key: url
            name: slack-alert-webhook-secret
          channel: '#alerts'
      route:
        groupBy: [alertname]
        receiver: slack
  writeConnectionSecretToRef:
    name: postgres-creds
----

> === Send E-Mail Alert through Alert-Manager and mailgun

Reviewer 2: ```suggestion
=== Send E-Mail Alert through Alertmanager and mailgun
```

VSHN has the option to send e-mails automatically through https://www.mailgun.com/[mailgun].
This can be leveraged to inform service users through e-mail without a lot of setup overhead.

In practice this can be implemented fairly similarly to forwarding alerts to Opsgenie.

> The composition will create an `AlertManagertConfig` resource in the database namespace that configures Alertmanager to send mails to the provided e-mail address.

Reviewer 3: ```suggestion
The composition will create an `AlertManagertConfig` resource in the database namespace that configures Alertmanager to send mails to the provided e-mail address.
```
The composition will create an `AlertManagerConfig` resource in the database namespace that configures Alertmanager to send mails to the provided e-mail address.

> The composition will create an `AlertManagertConfig` resource in the database namespace that configures alert-manager to send mails to the provided e-mail address.

Reviewer 2: ```suggestion
The composition will create an `AlertManagertConfig` resource in the database namespace that configures Alertmanager to send mails to the provided e-mail address.
```

> With such a system the database claim could look similar to the following, which would send all alerts as mails to `oncall@example.com`

Reviewer 2: ```suggestion
With such a system the database claim could look similar to the following, which would send all alerts as e-mails to `oncall@example.com`
```

[source,yaml]
----
apiVersion: vshn.appcat.vshn.io/v1
kind: VSHNPostgreSQL
metadata:
  name: pgsql-app1-prod
  namespace: prod-app
spec:
  parameters:
    alerting:
      email: oncall@example.com
  writeConnectionSecretToRef:
    name: postgres-creds
----

There are a few things we'll need to consider for such an implementation:

* This setup forces us to replicate a mailgun token to every database namespace which increases the potential attack surface.

> * We might need to think about verifying the e-mail before forwarding alerts to it.

Reviewer 2: Why do we need to verify e-mails?

Author: Not sure if we'll need to. But this setup will lead to us sending e-mails from our address to whatever address the user provides. Gives me a bad feeling, but maybe that's fine

Reviewer 4: As someone who receives a ton of unsolicited mails (like actual doctor appoints for other people...) on my private mail account I strongly agree to have some sort of mail verification in place.

This could be done through a custom controller.

== Decision

We implement both exposing the user-workload monitoring and automated e-mail alerts through mailgun.

> E-mail based alerting should be mandatory to ensure that we can prove that the service user was informed of the issue.

Reviewer: Not sure yet about that decision. At least with "should" we have some wiggle room for figuring out the exact implementation.

Reviewer: Perfect, thanks.

> E-mail based alerting should be strong encouraged, and might be mandatory in some cases, to ensure that we can prove that the service user was informed of the issue.

Reviewer 2: ```suggestion
E-mail based alerting should be strongly encouraged, and might be mandatory in some cases, to ensure that we can prove that the service user was informed of the issue.
```

=== Rationale

Manual intervention for capacity alerts generates a lot of unnecessary toil, as in most cases the engineer can't do anything but inform the service user.
That's why we believe that the task should be automated, and no engineer should be alerted for potential capacity issues.

Exposing the whole capabilities of the user-workload monitoring system is fairly simple to implement, while at the same time very versatile.
It allows our service users to send alerts exactly where they are needed.

On the other hand automatic e-mail alerts through mailgun has a much better UX for simple deployments.

> With this solution, service users don't need to handle any mail gateway or care about alert-manager configuration.

Reviewer 2: ```suggestion
With this solution, service users don't need to handle any mail gateway or care about Alertmanager configuration.
```

They simply provide an e-mail address, and they will be informed about capacity issues.

> A further advantage of sending alerts as e-mails through mailgun is, that we have a clear paper trail that the alert was sent and received by the service user.

Reviewer 2: Technically we can't really prove an alert was received. It might be discarded by the service users mail provider.

Author: I'm no e-mail expert, but can't we prove that their mail sever did not bounce the e-mail and at that point it's their fault if the e-mail didn't end up at the user?

Reviewer 4: We can proof that their server received a mail, what happens afterward is not in our power. Not sure if that's enough if this would become a legal matter...
